### ISSUE TITLE
'I am viewing a blog post' broken

### ISSUE THREAD

**Issue by stephenharris** [paulgibbs/behat-wordpress-extension#60]
## Expected behaviour

There are two errors, and a few improvements:

1. It seems like I should be able to invoke the step with any of the following:

```
I am viewing a blog post:
They are viewing a blog post:
I am viewing a post:
They are viewing a post:
```

2. The step when invoked creates the content and views the page.

3. It would be good to be able to view an existing post (by specifying a post title).

4. We should ideally retrieve the URL from the driver, rather than using `?p=` - this would implicitly check the rewrite rules are working (if present) and is more user-like behaviour.

## Current behaviour
1. Only the first two work. 
2. The step fails because post title / content were empty
3. This is not current possible.
4. We use `?p=...` to create the URL.

## Possible solution
1. The regex should be `^(?:I am|they are) viewing a(?: blog)? post:`
2. The `$post_data` array is actually an array of rows. Each row is the data of that post. We need to extract the first row (presumably). 
3. The regex could be set to `^(?:I am|they are) viewing a(?: blog)? post(?: "([^"]+)"|:)`
4. Introduce `getContentUrl($post_id)` to the driver.

## Steps to reproduce (for bugs)
1. Try using the step :)

I appreciate I'm including bugfixes and improvements for convenience, so I'll commit each of 1-4 separately so that they can be cherry-picked if we decide not to go with 3 & 4.

**Comment by stephenharris** [comment]
@paulgibbs What is the desired behaviour when we have multiple posts defined in this step?

```
Example: Given there are posts:
  | post_type  | post_title      | post_content | post_status |
  | page       | Test Post       | Hello World  | publish     |
  | page       | Some other page | Hello World  | publish     |
```


1. Ignore all but the first
2. Create all but view the first?
3. Create all but view the last?
4. Throw an exception
5. Other

**Comment by paulgibbs** [comment]
Oops!

1-4: agree. ✅

The `?p=x` thing was definitely a lazy shortcut. I like your suggestion that we'd also fail if the rewrite rules were broken, which is a good thing to catch.

Rather than something like `getContextUrl`, I'd suggest we follow CRUD naming concept. e.g. We already have `createContent` and `deleteContent`, let's add `getContent`. What that returns could be context-specific, or perhaps take a parameter that controls what value is returned (e.g. it could return an ID or a URL, for content).

**Comment by paulgibbs** [comment]
> What is the desired behaviour when we have multiple posts defined in this step?
Your example uses "Given there are posts", so I would expect all are created. None are viewed.

If you meant that example to be `I am viewing a blog post` which would fit the context of this issue :) -- good question. My first reaction says we should only create that first post (because the step is written singular, not plural), and then throw an exception. Not strongly opinionated about the exception bit, happy to go with what you or others think makes sense there.

**Comment by stephenharris** [comment]
Yes, I meant `I am viewing a blog post`.

Lets go with an exception. Failing silently might just lead to confusion. With an exception the problem is identified and easily rectified.

**Comment by stephenharris** [comment]
Regarding `getContentUrl` or similar,  this is a bit of a pain point for the `wp-cli` driver. Under `wp-api` I could request a `WP_Post` object. I had originally dismissed this idea as over-engineering, and maybe it is, but it's probably worth mentioning:

Ignoring the names for now. We could define `Content` interface, for example:
 
```
interface Content
 {
     public function id();
     public function title();
     public function author();
     public function type();
     public function url();
     public function content();
 }
```

Then, `Driver::getContent()` would have to return `Content`. The `wp-api` just needs to implement a wrapper around `WP_Post`. The `wp-cli` driver could front-load the data (we're kind of doing this with here, but using an array: https://github.com/paulgibbs/behat-wordpress-extension/blob/master/src/Driver/WpcliDriver.php#L273). Or we could 'lazy load it'. E.g.

```
WPCliContent implements Content
{
   protected $driver;
   protected $id;
   protected $url;
   function__construct( $id, WpcliDriver $driver ) {
      $this->id = $id;
      $this->driver = $driver;
   }

   public function url() {
       if ( ! $this->url ) {
          $this->url = $this->driver->wpcli( '... command to get url...' );
       }
       return $this->url;
   }
}
```

The `Driver`s would return only objects, and from there you can retrieve the data you need.

**Comment by paulgibbs** [comment]
RE: last comment, that does look a bit over-engineered. What you have done in your dev branch so far for the WP-CLI driver feels a more appropriate first try at this.

**Comment by stephenharris** [comment]
Fixed in #64
